### PR TITLE
JAMES 3897: Implement CrowdsecEhloHook

### PR DESCRIPTION
**Pull request by thanhbv200585**

### CODE REVIEW

**Comment by quantranhong1999** [comment]
[JAMES-3897: Implement CrowdsecHttpClient](https://github.com/apache/james-project/pull/1799/commits/12af37ef11df72ca64b4ce8d21029ef005b085e5) should be marked with the prefix [NO REVIEW] ...

File: third-party/crowdsec/docker-compose.yml [hunk 1]
```diff
@@ -10,6 +10,8 @@ services:
       - --generate-keystore
     volumes:
       - ./sample-configuration/logback.xml:/root/conf/logback.xml
+      - ./sample-configuration/extensions.properties:/root/conf/extensions.properties
+      - ./sample-configuration/smtpserver.xml:/root/conf/smtpserver.xml
```
Comment (quantranhong1999) [inline third-party/crowdsec/docker-compose.yml:14]: Missing  
- ./target/apache-james-crowdsec-jar-with-dependencies.jar:/root/extensions-jars/james-server-crowdsec.jar

?

File: third-party/crowdsec/pom.xml [hunk 2]
```diff
@@ -83,6 +84,10 @@
             <groupId>com.google.inject</groupId>
             <artifactId>guice</artifactId>
         </dependency>
+        <dependency>
+            <groupId>commons-net</groupId>
+            <artifactId>commons-net</artifactId>
```
Comment (chibenwa) [inline third-party/crowdsec/pom.xml:89]: Is this a test dependency?

File: third-party/crowdsec/src/test/java/org/apache/james/CrowdsecEhloHookTest.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    void givenIPNotBannedByCrowdsecDecisionIpRange() throws IOException, InterruptedException {
+        banIP("-r", "192.182.39.2/24");
+        int port = crowdsecExtension.getCrowdsecContainer().getMappedPort(CROWDSEC_PORT);
+        CrowdsecEhloHook ehloHook = new CrowdsecEhloHook(new CrowdsecClientConfiguration(new URL("http://localhost:" + port + "/v1"), CrowdsecClientConfiguration.DEFAULT_API_KEY));
+        SMTPSession session = new BaseFakeSMTPSession() {};
+
+        assertThat(ehloHook.doHelo(session, "localhost")).isEqualTo(HookResult.DECLINED);
+    }
+
+    private static void banIP(String type, String value) throws IOException, InterruptedException {
```
Comment (chibenwa) [inline third-party/crowdsec/src/test/java/org/apache/james/CrowdsecEhloHookTest.java:86]: `-i` and `-r` do not speak at all to me.

I suggest expliciting this...

Does this support longer options? eg `--range` ?

File: third-party/crowdsec/src/test/java/org/apache/james/CrowdsecEhloHookTest.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    @BeforeAll
+    static void setUp() throws IOException, InterruptedException {
+        crowdsecExtension.getCrowdsecContainer().execInContainer("cscli", "bouncer", "add", "bouncer", "-k", DEFAULT_API_KEY);
+    }
+
+    @Test
+    void givenIPBannedByCrowdsecDecisionIp() throws IOException, InterruptedException {
+        banIP("-i", "127.0.0.1");
+        int port = crowdsecExtension.getCrowdsecContainer().getMappedPort(CROWDSEC_PORT);
+        CrowdsecEhloHook ehloHook = new CrowdsecEhloHook(new CrowdsecClientConfiguration(new URL("http://localhost:" + port + "/v1"), CrowdsecClientConfiguration.DEFAULT_API_KEY));
```
Comment (chibenwa) [inline third-party/crowdsec/src/test/java/org/apache/james/CrowdsecEhloHookTest.java:50]: Put this in a `@BeforeEach` ?

File: third-party/crowdsec/src/test/java/org/apache/james/CrowdsecEhloHookTest.java (hunk 3 above)
Comment (thanhbv200585) [reply third-party/crowdsec/src/test/java/org/apache/james/CrowdsecEhloHookTest.java:86]: Yep, I will change it

**Comment by thanhbv200585** [comment]
[Crowdsec EhloHook Demo.webm](https://github.com/apache/james-project/assets/87489927/ecae16a6-8064-4f3a-b6d9-f1b3fbdb75c4)

File: third-party/crowdsec/pom.xml [hunk 5]
```diff
@@ -40,6 +40,19 @@
             <type>test-jar</type>
             <scope>test</scope>
         </dependency>
```
Comment (chibenwa) [inline third-party/crowdsec/pom.xml:42]: Please use `scope provided` for dependencies shiped as part of the james container in order to reduce the size of your extension.

File: third-party/crowdsec/pom.xml [hunk 6]
```diff
@@ -57,15 +80,39 @@
...
             <artifactId>james-server-testing</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.apache.james.protocols</groupId>
+            <artifactId>protocols-smtp</artifactId>
+        </dependency>
         <dependency>
             <groupId>org.assertj</groupId>
             <artifactId>assertj-core</artifactId>
```
Comment (chibenwa) [inline third-party/crowdsec/pom.xml:118]: Jupiter-junit -> use the version provided by Apache James

**Review by quantranhong1999 — COMMENTED** [review]
Look good to me overall.

Can you add a readme on how to use the CrowdSec extension with James? (may have a look at Rspamd readme)

File: third-party/crowdsec/pom.xml (hunk 5 above)
Comment (chibenwa) [reply third-party/crowdsec/pom.xml:42]: Keep test on that one!

**Review by Arsnael — COMMENTED** [review]
Read it, nothing to add. Good job overall :)

**Review by Arsnael — APPROVED** [review]

File: third-party/crowdsec/README.md [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+- For running docker-compose, first compile this project
+
+```
+mvn clean install -DskipTests
+```
+then run it: `docker-compose up`
+
+## Additional webadmin endpoints
+
+### Check IP blocked by Crowdsec (decisions)
```
Comment (quantranhong1999) [inline third-party/crowdsec/README.md:39]: This is misleading. These are CrowdSec endpoints, not (James) additional webadmin endpoints.

Maybe just put a link to CrowdSec endpoints documentation, or change the way you paraphrase.

File: third-party/crowdsec/README.md [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+Crowdsec will expose port 8080 for queries to get the list of IP addresses blocked (get decisions)
+
+```bash
+curl -XGET http://localhost:8082/v1/decisions -H "X-Api-Key: default_api_key" -H 'accept: application/json' | jq .
```
Comment (quantranhong1999) [inline third-party/crowdsec/README.md:44]: 8080 then below 8082 (I know the exposed port changed, but still misleading).

BTW CrowdSec just needs to stay behind James, no need to expose CrowdSec service to host machine, no?

**Comment by quantranhong1999** [comment]
rebase mistake. Please fetch and rebase on top of the newest `apache/master` everytime you rebase.

**Comment by Arsnael** [comment]
Can rebase :)

**Review by quantranhong1999 — COMMENTED** [review]
Some minor points left. But good job :)

File: third-party/crowdsec/src/main/java/org/apache/james/CrowdsecEhloHook.java [hunk 9]
```diff
@@ -31,23 +31,19 @@
...
 
     @Inject
     public CrowdsecEhloHook(CrowdsecClientConfiguration configuration) {
-        this.crowdsecClientConfiguration = configuration;
+        this.crowdsecHttpClient = new CrowdsecHttpClient(configuration);
     }
 
     @Override
     public HookResult doHelo(SMTPSession session, String helo) {
         String ip = session.getRemoteAddress().toString().split("/")[1].split(":")[0];
```
Comment (chibenwa) [inline third-party/crowdsec/src/main/java/org/apache/james/CrowdsecEhloHook.java:44]: `session.getRemoteAddress().getAddress().getHostAddress();`

**Comment by Arsnael** [comment]
LGTM, just one more fix :)

**Review by Arsnael — APPROVED** [review]

**Review by chibenwa — APPROVED** [review]

**Pull request merged**